DQM/TrackingMonitor/src/dEdxHitAnalyzer.cc

```diff
@@ -0,0 +1,223 @@
+/*
+ *  See header file for a description of this class.
+ *
+ *  \author Loic Quertenmont 
+ */
+#include "DQM/TrackingMonitor/interface/dEdxHitAnalyzer.h"
+
+#include "TrackingTools/TransientTrack/interface/TransientTrackBuilder.h"
+#include "TrackingTools/Records/interface/TransientTrackRecord.h"
+#include "TrackingTools/TransientTrack/interface/TransientTrack.h"
+#include "MagneticField/Engine/interface/MagneticField.h"
+
+#include "CommonTools/TriggerUtils/interface/GenericTriggerEventFlag.h"
+
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "FWCore/Utilities/interface/InputTag.h"
+#include "DQMServices/Core/interface/DQMStore.h"
+
+#include <string>
+#include "TMath.h"
+
+dEdxHitAnalyzer::dEdxHitAnalyzer(const edm::ParameterSet& iConfig) 
+  : dqmStore_( edm::Service<DQMStore>().operator->() )
+  , fullconf_( iConfig )
+  , conf_    (fullconf_.getParameter<edm::ParameterSet>("dEdxParameters") )
+  , doAllPlots_  ( conf_.getParameter<bool>("doAllPlots") )
+  , doDeDxPlots_ ( conf_.getParameter<bool>("doDeDxPlots") )    
+  , genTriggerEventFlag_( new GenericTriggerEventFlag(conf_,consumesCollector()) )
+{
+
+  trackInputTag_ = edm::InputTag(conf_.getParameter<std::string>("TracksForDeDx") );
+  trackToken_ = consumes<reco::TrackCollection>(trackInputTag_);
+
+  dEdxInputList_ = conf_.getParameter<std::vector<std::string> >("deDxHitProducers");
+  for (auto const& tag : dEdxInputList_) {
+    dEdxTokenList_.push_back(consumes<reco::DeDxHitInfoAss>(edm::InputTag(tag) ) );
+  }
+}
+
+dEdxHitAnalyzer::~dEdxHitAnalyzer() 
+{ 
+
+  if (genTriggerEventFlag_)      delete genTriggerEventFlag_;
+}
+
+// ------------ method called once each job just after ending the event loop  ------------
+void 
+dEdxHitAnalyzer::endJob() 
+{
+    bool outputMEsInRootFile   = conf_.getParameter<bool>("OutputMEsInRootFile");
+    std::string outputFileName = conf_.getParameter<std::string>("OutputFileName");
+    if(outputMEsInRootFile)
+    {
+      dqmStore_->showDirStructure();
+      dqmStore_->save(outputFileName);
+    }
+}
+
+/*
+// -- BeginRun
+//---------------------------------------------------------------------------------//
+void dEdxHitAnalyzer::beginRun(const edm::Run& iRun, const edm::EventSetup& iSetup)
+{
+ 
+  // Initialize the GenericTriggerEventFlag
+  if ( genTriggerEventFlag_->on() ) genTriggerEventFlag_->initRun( iRun, iSetup );
+}
+*/
+
+void dEdxHitAnalyzer::bookHistograms(DQMStore::IBooker & ibooker,
```

Comment:
please move everything which is not pure booking outside the bookhistograms method. you can use the dqmBeginRun instead.
the getParameter can be done in the ctor instead.